File: A1/sleep-sort/sleep-sort.py
"""
    sleep sort in Python
    from https://gist.github.com/ssirai/4115014
"""

import thread
import sys
from time import sleep
from random import shuffle
from threading import Thread
from numpy import arange
import psutil

r = []
class worker(Thread):
    def __init__(self, t):
        Thread.__init__(self)
        self.t = t

    def run(self):
        global r
        sleep(self.t)
        self.t = "%0.2f" % self.t
        r.append(self.t)
        print self.t
        sys.stdout.flush()
        
class Monitor(Thread):
    def run(self):
        iterations = 10
        CPUusage = []
        VMusage = []
        for i in range(iterations):
            CPU = psutil.cpu_percent(0.2, False)
            print "CPU usage:", CPU
            CPUusage.append(CPU)
            VM = psutil.virtual_memory()
            print VM
            VMusage.append(VM[2])
            sleep(1)
        CPUavg = sum(CPUusage)/float(len(CPUusage))
        print "Average CPU usage:", CPUavg
        VMavg = sum(VMusage)/float(len(VMusage))
        print "Average VM usage:", VMavg
        #thread.interrupt_main()

Monitor().start()
sleep(1)

high = float(sys.argv[1])
step = float(sys.argv[2])
xs = [i for i in arange(0, high, step)]
shuffle(xs)

xs = map(worker, xs)
for x in xs:
    x.start()

for x in xs:
    x.join()

print "Sorted:", all(r[i] <= r[i+1] for i in xrange(len(r)-1))

File: A1/sleep-sort/sleep-sort.go
/*  sleep sort in Go
*   from https://github.com/arpitbbhayani/go-sleep-sort/blob/master/sleepsort.go
*/

package main

import (
	"fmt"
	"os"
	"strconv"
	"sync"
	"time"
	//"github.com/shirou/gopsutil/cpu"
	//"github.com/shirou/gopsutil/mem"
)

var max = 10.0
var step = 1.0

// prints a number of sleeping for n seconds
func sleepAndPrint(x float64, wg *sync.WaitGroup) {
	defer wg.Done()

	// Sleeping for time proportional to value
	time.Sleep(time.Duration(x) * 1000.0 * time.Millisecond)

	// Printing the value
	fmt.Printf("%f\n", x)
}

// Sorts given integer slice using sleep sort
func Sort(numbers []float64) {
	var wg sync.WaitGroup

	// Creating wait group that waits of len(numbers) of go routines to finish
	wg.Add(len(numbers))

	for _, x := range numbers {
		// Spinning a Go routine
		go sleepAndPrint(x, &wg)
	}

	// Waiting for all go routines to finish
	wg.Wait()
}

/*
func monitor () {
    iterations := 10
    CPUusage := 0.0
    VMusage := 0.0
    for i := 0; i < iterations; i++ {
      c, err := cpu.Percent(time.Duration(200) * time.Millisecond, false)
      m, err := mem.VirtualMemory()
      if err == nil {
        cpu := c[0]
        vm := m.UsedPercent
        CPUusage += cpu
        VMusage += vm
        fmt.Printf("CPU usage: %f\n", cpu)
        fmt.Printf("%v\n", m)
      }
      time.Sleep(time.Second)
    }
    CPUavg := CPUusage / float64(iterations)
    fmt.Printf("Average CPU usage: %f \n", CPUavg)
    VMavg := VMusage / float64(iterations)
    fmt.Printf("Average VM usage: %f \n", VMavg)
		os.Exit(0);
}*/

func main() {
	//go monitor()
	//time.Sleep(time.Second)

	args := os.Args[1:]

	max, err := strconv.ParseFloat(args[0], 64)
	if err != nil {
		panic(err)
	}

	step, err := strconv.ParseFloat(args[1], 64)
	if err != nil {
		panic(err)
	}

	var numbers = []float64{}
	for xi := 0.0; xi < max; xi += step {
		numbers = append(numbers, xi)
	}

	Sort(numbers)
}
